modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/MonadLogic.kt

```diff
@@ -0,0 +1,45 @@
+package arrow.typeclasses
+
+import arrow.Kind
+import arrow.core.ForOption
+import arrow.core.Option
+import arrow.core.Tuple2
+import arrow.core.fix
+import arrow.core.identity
+
+interface MonadLogic<F> : MonadPlus<F> {
+
+  fun <A> Kind<F, A>.msplit(): Kind<F, Option<Tuple2<Kind<F, A>, A>>>
+
+  fun <A> Kind<F, A>.interleave(fa: Kind<F, A>): Kind<F, A> =
+    this.msplit().flatMap { option ->
+      option.map { (fa, a) ->
+        just(a).mplus(fa.interleave(fa))
+      }.fold({ fa }, ::identity)
+    }
+
+  fun <A, B> Kind<F, A>.fairConjunction(ffa: (A) -> Kind<F, B>): Kind<F, B> =
+    msplit().flatMap { option ->
+      option.fold({ mzero<Tuple2<Kind<F, A>, A>>() }, { just(it) })
+    }.flatMap { (fa, a) ->
+      ffa(a).interleave(fa.flatMap(ffa))
+    }
+
+  fun <A, B> Kind<F, A>.ifte(fb: Kind<F, B>, ffa: (A) -> Kind<F, B>): Kind<F, B> =
```

Comment:
I'm okay with longer names. `ifThen` if that's what this does, or something more descriptive 🙏